core/src/sql/number.rs

```diff
@@ -615,42 +617,102 @@ impl Ord for Number {
 			// ------------------------------
 			(Number::Float(v), Number::Decimal(w)) => {
-				if v > &Decimal::MAX
-					.to_f64()
-					.expect("Decimal::MAX should be able to be converted to f64")
-				{
-					Ordering::Greater
-				} else if v < &Decimal::MIN
-					.to_f64()
-					.expect("Decimal::MIN should be able to be converted to f64")
-				{
-					Ordering::Less
-				} else if let Some(vd) = Decimal::from_f64_retain(*v) {
-					match (vd.cmp(w), v.fract() == 0.0, w.fract() == Decimal::ZERO) {
-						// both non-integers so we order float first
-						(Ordering::Equal, false, false) => Ordering::Less,
-						// both have equal int parts, but w has larger magnitude
-						(Ordering::Equal, false, true) => match v.is_sign_positive() {
-							true => Ordering::Greater,
-							false => Ordering::Less,
-						},
-						(Ordering::Equal, true, false) => match v.is_sign_positive() {
-							true => Ordering::Less,
-							false => Ordering::Greater,
-						},
-						// Both are integers and equal
-						(Ordering::Equal, true, true) => Ordering::Equal,
-						(o @ Ordering::Less | o @ Ordering::Greater, _, _) => o,
+				// Compare fractional parts of the float and decimal.
+				macro_rules! compare_fractions {
+					($l:ident, $r:ident) => {
+						match ($l == 0.0, $r == Decimal::ZERO) {
+							// If both numbers are zero, these are equal.
+							(true, true) => {
+								return Ordering::Equal;
+							}
+							// If only the float is zero, check the decimal's sign.
+							(true, false) => {
+								return greater!($r).reverse();
+							}
+							// If only the decimal is zero, check the float's sign.
+							(false, true) => {
+								return greater!($l);
+							}
+							// If neither is zero, continue checking the rest of the digits.
+							(false, false) => {
+								continue;
+							}
+						}
+					};
+				}
+				// If the float is not finite, we don't need to compare it to the decimal
+				if !v.is_finite() {
+					return greater!(v);
+				}
+				// Cast int to i128 to avoid saturating.
+				let l = *v as i128;
+				// Cast the integer-part of the decimal to i128.
+				let Ok(r) = i128::try_from(*w) else {
+					return greater!(w).reverse();
+				};
+				// Compare both integer parts.
+				match l.cmp(&r) {
+					// If the integer parts are equal then we need to compare the fractional parts.
+					Ordering::Equal => {
+						// We can't compare the fractional parts of floats with decimals reliably.
+						// Instead, we need to compare them as integers. To do this, we need to
+						// multiply the fraction with a number large enough to move some digits
+						// to the integer part of the float or decimal. The number should fit in
+						// an f64 and be a multiple of 2. Since we may need to do this repeatedly
+						// it helps if the number is as big as possible to reduce the number of
+						// iterations needed.
+						const SAFE_MULTIPLIER: i64 = 9_007_199_254_740_000;
```

Comment:
Where does this number come from? wouldn't it make more sense for it to be a power of 2?

Reply:
It's a magic number 🤣 It's 2^53 with the last 12 bits zeroed out. This is so that we can actually shit the fractional part towards zero with each iteration, which would otherwise not be possible. Without this property, equal fractions would end up in an endless loop.

Reply:
I have updated the multiplier to a more efficient one.

Reply:
The new multiplier fails the following test:

```
assertion `left != right` failed: Float(0.763730645766375) Float(0.7637306457663751) Decimal(0.763730645766375)
  left: Float(0.763730645766375)
 right: Decimal(0.763730645766375)
```

I have marked the PR as draft while I run some more tests.

Reply:
I reverted the new multiplier. Let's just use the magic number for now 😆 For most use-cases, it's just as efficient anyway.